add duckdb and motherduck destination connector

krish-adi opened the pull request:
This PR adds support for DuckDB and MotherDuck destination connector.

potter-potter commented on requirements/connectors/duckdb.in:
```diff
@@ -0,0 +1,3 @@
+-r ../common/base.in
+
+duckdb==1.1.0
```
This is great for the requirements. Could you please add a line in this file:
https://github.com/Unstructured-IO/unstructured-ingest/blob/b3add2abf347b5e21121bf01626f8b4e9e3ec9e1/setup.py#L113

It is what gives us the ability for the user to run `pip install "unstructured-ingest[duckdb]"

krish-adi replied:
@potter-potter would this : https://github.com/krish-adi/unstructured-ingest/blob/625797062d3b151d896d1a4a4cea39dbb06b889e/setup.py#L97  work?

potter-potter replied:
I think that is it. I'll let you know if theres more. (Never remember this part. I always just search for `pinecone` and see where we added them.

potter-potter commented on test_e2e/env_setup/duckdb/create-duckdb-schema.py:
```diff
@@ -0,0 +1,13 @@
+from pathlib import Path
+
+if __name__ == "__main__":
+    connection = duckdb.connect(sys.argv[1])
```
I'm getting an error here. Can duckdb really connect to a completely empty .db file. if so, cool!

`duckdb.duckdb.IOException: IO Error: The file "/Users/blabla/test_e2e/elements.db" exists, but it is not a valid DuckDB database file!`

potter-potter replied:
Looks like you just need to remove the `touch` in the other file.

krish-adi replied:
that makes sense. the duckdb client creates the database file when a database does not exists.

potter-potter commented on test_e2e/env_setup/duckdb/create-duckdb-instance.sh:
```diff
@@ -0,0 +1,14 @@
+DATABASE_FILE_PATH=$2
+
+# Create the DuckDB instance
+touch "$DATABASE_FILE_PATH"
```
This line caused problems for me. Worked better when commented out.

krish-adi replied:
makes sense.

potter-potter commented on unstructured_ingest/v2/processes/connectors/duckdb.py:
```diff
@@ -0,0 +1,238 @@
+    "points",
+    "url",
+    "version",
+    "date_created",
```
These columns come in from partitioning as int64 and float64:


date_created                                             1728685124
date_modified                                            1728685124
date_processed                                    1728929563.441951


So the SQL insert is having trouble with them. Probably converting them to timestamp

krish-adi replied:
@potter-potter is this different from `last_modified` ? because when this line https://github.com/krish-adi/unstructured-ingest/blob/624371416d9661b1f5a81a2af0a2fa216fdd598b/unstructured_ingest/v2/processes/connectors/duckdb.py#L220 is executed, the df_elements stores the last_modified as this `2024-09-27T05:10:47`. 

is there a file or an option i can turn on to be able to test this out locally? the script i have to test locally never adds this metadata.

potter-potter replied:
Yeah. I can't remember why we get timestamps for last_modified and unix epoch numbers or whatever for the other ones.

Just try to intercept the DataFrame right before it writes and check out the first row.

potter-potter replied:
I will look to see if we even need them converted to timestamps, or if we just leave them as numbers when I'm back online tomorrow.

krish-adi replied:
this change: https://github.com/krish-adi/unstructured-ingest/blob/75774b7b5bcd9214302a77d0bc229d33c2bc75b3/test_e2e/env_setup/duckdb/create-duckdb-schema.sql#L13-L15

should hopefully resolve the insert conflicts.

krish-adi replied:
hey @potter-potter let me know if these changes are a possible workaround....

potter-potter commented on test_e2e/dest/duckdb.sh:
```diff
@@ -0,0 +1,53 @@
+  --db-type "$DATABASE_TYPE" \
+  --database "$DB_PATH"
+
+"$SCRIPT_DIR"/env_setup/duckdb/test_outputs.py \
```
this should be test-outputs.py not underscore

potter-potter commented on test_e2e/dest/duckdb.sh:
```diff
@@ -0,0 +1,53 @@
+source "$SCRIPT_DIR"/cleanup.sh
+function cleanup {
+  # Local file cleanup
+  cleanup_dir "$WORK_DIR"
```
Comment these 2 lines out so you can see the files that are created in this test. For the metadata debugging.

potter-potter replied:
At least its part of how I debug metadata etc...

potter-potter commented:
This is looking great! I think we are almost there.

rbiseck3 commented on CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+## 0.0.27
```
Can this version bump be the next `-dev` iteration rather than a new version? This lets us control what gets bundled into a single release.

rbiseck3 commented on requirements/connectors/duckdb.in:
```diff
@@ -0,0 +1,3 @@
+-r ../common/base.in
```
Don't want to bundle in the base dependencies into the connector specific ones. 
```suggestion
-c ../common/constraints.txt
```

rbiseck3 commented on test_e2e/dest/duckdb.sh:
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env bash
```
We've recently updated to add integration tests rather than e2e tests for new connectors, take a look at the s3 example: [test_s3.py](https://github.com/Unstructured-IO/unstructured-ingest/blob/main/test/integration/connectors/test_s3.py). Could we drop this file and add a new integration test instead? That should help isolate just the code being added in this PR.

rbiseck3 commented on unstructured_ingest/v2/processes/connectors/duckdb.py:
```diff
@@ -0,0 +1,238 @@
+MOTHERDUCK_DB = "motherduck"
+
+
+class MotherDuckAccessConfig(AccessConfig):
```
Can we support these as 2 distinct connectors that are based on common base classes? Take a look at this PR I recently opened to update the sql connector that did something similar to support distinct connectors: https://github.com/Unstructured-IO/unstructured-ingest/pull/181

rbiseck3 commented:
I've added the `needs edits` label to this PR for now. Feel free to remove it once it's ready to have someone rereview it.

potter-potter commented:
@krish-adi We are in the middle of a big code push this week. I will help you with Roman's comments for this PR at the beginning of next week.

krish-adi commented:
hey @potter-potter good luck with the release. I'll follow up on the fixes requested above, and you can take a look next week.

krish-adi commented:
@potter-potter , hope the release went well! I am wondering what changes need to be made here to be compatible with the newer version of the library? Could you review it and let me know, and I could get the changers in by this week.

potter-potter commented:
@rbiseck3 I am assuming your comments on this PR are still accurate? Were there any async or multiprocessing changes that would need to be looked at? (thats the only think I can think of that might have changed in the last 3 weeks)

@krish-adi Thanks. Release went well. But of course we are slammed again for another sprint. :) I would say that Roman's changes are still accurate. And would help take this over the finish line.

potter-potter commented:
Also there should be some good examples of splitting out the 2 distinct connectors now based on authentication in the repo now. Like for databricks_volumes. This will help make the connector easier to integrate in our Platform UI.

potter-potter commented:
@krish-adi please switch the tag to  `ready for review` when you feel its ready

krish-adi commented:
to test against motherduck, you would require a MOTHERDUCK_TOKEN. you can get a free account from motherduck.com ([docs on this here](https://motherduck.com/docs/key-tasks/authenticating-and-connecting-to-motherduck/authenticating-to-motherduck/))and the free limits would be enough for testing the integration. 

how often would this run in your CI ?

krish-adi commented:
hey folks, any update on this one?

potter-potter commented:
> hey folks, any update on this one?

thanks for the bump. I'll look at it this evening. We have been turbo slammed. (i know seems like always right now...)

potter-potter commented:
Hey. This a really cool. Motherduck is a very nice experience. Sort of a lightweight databricks. A pleasure to use.

The connector works well. Can you put the motherduck.py and duckdb.py into a folder called duckdb?

Can you address this warning I am getting in the tests:
`UserWarning: Field name "schema" in "DuckDBConnectionConfig" shadows an attribute in parent "ConnectionConfig"`

krish-adi commented:
Hey @potter-potter ! great to hear that :) thanks for suggestions. I've moved the motherduck and duckdb connectors and their tests into duckdb folder. I've also renamed the schema to db_schema to not conflict with the parent.

krish-adi commented:
Hey @potter-potter and @rbiseck3 , any other changes required for this connector?

krish-adi commented:
Hey @potter-potter let me know when this is released and lands on your platform. We could do a shoutout on socials.

potter-potter commented:
@krish-adi Closing this because it is now in the latest release via the other PR. Thank you!
